src/3d/symbols/qgspolygon3dsymbol_p.cpp

```diff
@@ -224,4 +254,55 @@ Qt3DRender::QGeometryRenderer *QgsPolygon3DSymbolEntityNode::renderer( const Qgs
   return renderer;
 }
 
+Qt3DRender::QGeometryRenderer *QgsPolygon3DSymbolEntityNode::renderer( const Qgs3DMapSettings &map,
+    const QgsPolygon3DSymbol &symbol,
+    const QgsMeshLayer *layer )
+{
+  QgsPointXY origin( map.origin().x(), map.origin().y() );
+  QList<QgsPolygon *> polygons;
+  QList<QgsFeatureId> fids;
+  QList<float> extrusionHeightPerPolygon;  // will stay empty if not needed per polygon
+
+  QgsExpressionContext ctx( _expressionContext3D() );
+  const QgsPropertyCollection &ddp = symbol.dataDefinedProperties();
+  bool hasDDHeight = ddp.isActive( QgsAbstract3DSymbol::PropertyHeight );
+  bool hasDDExtrusion = ddp.isActive( QgsAbstract3DSymbol::PropertyExtrusionHeight );
+  float height = symbol.height();
+  float extrusionHeight = symbol.extrusionHeight();
+  if ( hasDDHeight )
+    height = ddp.valueAsDouble( QgsAbstract3DSymbol::PropertyHeight, ctx, height );
+  if ( hasDDExtrusion )
+    extrusionHeight = ddp.valueAsDouble( QgsAbstract3DSymbol::PropertyExtrusionHeight, ctx, extrusionHeight );
+
+  const QgsTriangularMesh *mesh = layer->triangularMesh();
+  if ( layer && mesh )
+  {
+    const QVector<QgsMeshFace> &triangles = mesh->triangles();
+    const QVector<QgsMeshVertex> &vertices = mesh->vertices();
+    for ( int i = 0; i < triangles.size(); ++i )
+    {
+      const QgsMeshFace &triangle = triangles.at( i );
+      Q_ASSERT( triangle.size() == 3 );
+      std::unique_ptr< QgsPolygon > polygon = QgsMeshUtils::toPolygon( triangle, vertices );
+      Qgs3DUtils::clampAltitudes( polygon.get(), symbol.altitudeClamping(), symbol.altitudeBinding(), height, map );
+      polygons.append( polygon.release() );
+      fids.append( i );
+      if ( hasDDExtrusion )
+        extrusionHeightPerPolygon.append( extrusionHeight );
```

Comment:
I am not sure if extrusion makes sense for mesh - I would prefer not to use it and not even show it in the GUI